### PR TITLE
DO NOT MERGE: Federation notifications processing

### PR DESCRIPTION
**Pull request by diegomasini**
As described in RSKIP 68: https://github.com/rsksmart/RSKIPs/blob/federation_notifications/IPs/RSKIP68.md this PR includes changes to allow RSK Nodes to process Federation Notifications generated by Federated Nodes to detect potential attacks to the network.

**IMPORTANT:** This implementation is meant to be a reference implementation. As such, do not merge it as it is and feel free to modify it as you see appropriate.

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java [hunk 1]
```diff
@@ -321,6 +350,35 @@ private void processBlockMessage(@Nonnull final MessageChannel sender, @Nonnull
         tryRelayBlock(sender, block, result);
         recordEvent(sender, EventType.VALID_BLOCK);
         Metrics.processBlockMessage("finish", block, sender.getPeerNodeID());
+
+        // If still synchronizing then return
+        if (blockProcessor.hasBetterBlockToSync()) {
+            return;
+        }
+
+        // If we are a valid source for federation notifications then generate one
```
Comment (amendelzon) [inline rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: Why would this be here? Doesn't the federate already generate the notifications?

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java [hunk 2]
```diff
@@ -321,6 +350,35 @@ private void processBlockMessage(@Nonnull final MessageChannel sender, @Nonnull
...
+        if (federationNotificationSource != null) {
+            federationNotificationSource.generateNotification();
+        }
+
+        // Check if the Federation disappeared for a long time (Federation eclipsed).
+        federationNotificationProcessor.checkIfFederationWasEclipsed();
```
Comment (amendelzon) [inline rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: I would delegate this call to a more specific component, unless there's a good reason to do it here.

File: rskj-core/src/main/java/co/rsk/net/messages/MessageType.java [hunk 3]
```diff
@@ -243,6 +250,48 @@ public Message createMessage(RLPList list) {
...
+                long blockNumber = bytesToLong(rlpConfirmation.get(0).getRLPData());
+                Keccak256 blockHash = new Keccak256(rlpConfirmation.get(1).getRLPData());
+
+                notification.addConfirmation(blockNumber, blockHash);
+            }
+
+            return notification;
+        }
+
+        private long bytesToLong(byte[] bytes) {
```
Comment (amendelzon) [inline rskj-core/src/main/java/co/rsk/net/messages/MessageType.java]: Minor detail: no need for this, can use `BigInteger` or similar.

File: rskj-core/src/main/java/co/rsk/net/notifications/FederationNotificationSource.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.net.notifications;
+
+public interface FederationNotificationSource {
```
Comment (amendelzon) [inline rskj-core/src/main/java/co/rsk/net/notifications/FederationNotificationSource.java]: I'd probably move this and its implementation to the Federate node side, unless there's a good reason not to.

**Comment by tinchou** [comment]
Closing for now, feel free to open it again any time.

**Pull request closed without merging**